src/pages/filter/index.tsx

```diff
@@ -0,0 +1,258 @@
+import React, { useState, useMemo } from 'react';
+import { Table, Button } from 'semantic-ui-react';
+import { Link } from 'react-router-dom';
+import { Header, Segment } from 'semantic-ui-react';
+import MonacoEditor, { Monaco } from '@monaco-editor/react';
+import Pagination from '@/components/Pagination';
+import { Contest, OIer, School } from '@/libs/OIerDb';
+import usePartialSearchParams from '@/utils/usePartialSearchParams';
+import styles from './index.module.less';
+import fixContestName from '@/utils/fixContestName';
+import fixChineseSpace from '@/utils/fixChineseSpace';
+import jsDoc from './OIerDb.js?raw';
+
+const FilterWithIDE: React.FC = () => {
+  const [searchParams] = usePartialSearchParams();
+
+  const page = Number(searchParams.get('page') || 1);
+  const perPage = 30;
+
+  const handleEditorBeforeMount = (monaco: Monaco) => {
+    monaco.languages.typescript.javascriptDefaults.addExtraLib(jsDoc);
+  };
+
+  const STORAGE_KEY = 'monaco-editor-content';
+
+  const [filterCode, setFilterCode] = useState(
+    localStorage.getItem(STORAGE_KEY) ||
+      `
+/**
+ * @param {OIer[]} oiers
+ * @param {School[]} schools
+ * @param {Contest[]} contests
+ */
+function filter(oiers, schools, contests) {
+  return oiers.filter((oier) => {
+    return oier.ccf_level === 10;
+  });
+}
+
+module.exports = filter;      
+      `.trim() + '\n'
+  );
+
+  let saveTimeout: NodeJS.Timeout;
+  const handleEditorChange = (value: string | undefined) => {
+    setFilterCode(value);
+    clearTimeout(saveTimeout);
+    saveTimeout = setTimeout(() => {
+      if (value != null) {
+        localStorage.setItem(STORAGE_KEY, value);
+      }
+    }, 300);
+  };
+
+  const [filterError, setFilterError] = useState<string | null>(null);
+  const [activeFilter, setActiveFilter] = useState<string>(filterCode);
+
+  function loadAsModule(code: string) {
+    const module = { exports: {} };
+    const func = new Function('module', code);
+    func(module);
+    return module.exports;
+  }
+
+  const filteredData = useMemo(() => {
+    setFilterError(null);
+
+    if (!activeFilter.trim()) {
+      return { type: 'OIer', result: OIerDb.oiers };
+    }
+
+    try {
+      const { oiers, schools, contests } = OIerDb;
+
+      type LoadModuleFunction = (
+        oiers: OIer[],
+        schools: School[],
+        contests: Contest[]
+      ) => any; // eslint-disable-line @typescript-eslint/no-explicit-any
+
+      const result = (loadAsModule(activeFilter) as LoadModuleFunction)(
+        oiers,
+        schools,
+        contests.reverse()
```

Comment:
```suggestion
        contests.toReversed()
```
反正 `reverse()` 肯定是不对的，它会修改原数组，按照你的语义可以改成 `toReversed()`。不过可以传整个 OIerDb 对象，随便你了